java-checks/src/main/java/org/sonar/java/checks/UnconditionalJumpStatementCheck.java

```diff
@@ -0,0 +1,166 @@
+      reportIssue(jumpKeyword, String.format("Remove this \"%s\" statement or make it conditional.", jumpKeyword.text()));
+    }
+  }
+
+  private static boolean canExecuteMoreThanOnce(Tree loopTree) {
+    CFG cfg = getCFG(loopTree);
+    return cfg.blocks().stream().anyMatch(block -> loopTree.equals(block.terminator()) && hasPredecessorInLoopBody(block.predecessors(), loopTree));
+  }
+
+  private static boolean hasPredecessorInLoopBody(Set<Block> predecessors, Tree loop) {
+    for (CFG.Block predecessor : predecessors) {
+      List<Tree> predecessorElements = predecessor.elements();
+      if (!predecessorElements.isEmpty()) {
+        Tree predecessorLastElement = predecessorElements.get(predecessorElements.size() - 1);
+
+        if (isForStatementInitializer(predecessorLastElement, loop)) {
+          continue;
+        }
+
+        if (isForStatementUpdate(predecessorLastElement, loop)) {
+          // there is a way to reach the update
+          return !predecessor.predecessors().isEmpty();
+        }
+
+        if (isDescendant(predecessorLastElement, loop)) {
+          return true;
+        }
+      } else {
+        return hasPredecessorInLoopBody(predecessor.predecessors(), loop);
+      }
+    }
+    return false;
+  }
+
+  private static boolean isForStatementInitializer(Tree lastElement, Tree loop) {
+    if (loop.is(Tree.Kind.FOR_STATEMENT)) {
+      ListTree<StatementTree> initializer = ((ForStatementTree) loop).initializer();
+      if (!initializer.isEmpty()) {
+        return isDescendant(lastElement, initializer);
+      }
+    } else if (loop.is(Tree.Kind.FOR_EACH_STATEMENT)) {
+      return isDescendant(lastElement, ((ForEachStatement) loop).expression());
+    }
+    return false;
+  }
+
+  private static boolean isForStatementUpdate(Tree lastElement, Tree loop) {
+    if (loop.is(Tree.Kind.FOR_STATEMENT)) {
+      ListTree<StatementTree> update = ((ForStatementTree) loop).update();
+      if (!update.isEmpty()) {
+        return isDescendant(lastElement, update);
+      }
+    }
+    return false;
+  }
+
+  private static boolean isDescendant(Tree descendant, Tree target) {
+    Tree parent = descendant;
+    while (parent != null) {
+      if (parent.equals(target)) {
+        return true;
+      }
+      parent = parent.parent();
+    }
+    return false;
+  }
+
+  private static SyntaxToken jumpKeyword(Tree jumpStatement) {
+    if (jumpStatement.is(Tree.Kind.BREAK_STATEMENT)) {
+      return ((BreakStatementTree) jumpStatement).breakKeyword();
+    } else if (jumpStatement.is(Tree.Kind.CONTINUE_STATEMENT)) {
+      return ((ContinueStatementTree) jumpStatement).continueKeyword();
+    } else if (jumpStatement.is(Tree.Kind.RETURN_STATEMENT)) {
+      return ((ReturnStatementTree) jumpStatement).returnKeyword();
+    } else {
+      return ((ThrowStatementTree) jumpStatement).throwKeyword();
+    }
+  }
+
+  private static CFG getCFG(Tree loop) {
+    Tree currentTree = loop;
+    do {
```

Comment:
Wondering why you need to get back to the parents ? couldn't you consider the loop passed in tree as a list of statement and construct CFG solely for this one ? If i'm correct this should be sufficient.